### PR TITLE
expression: fix that str_to_date returns different result on no zero date sql mode.

### LINKED ISSUE #39146: `str_to_date` return different result on no zero date sql mode
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
mysql> CREATE TABLE `sun` ( `dest` varchar(10) DEFAULT NULL ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;
Query OK, 0 rows affected (0.17 sec)

mysql> insert into sun values('20231020');
Query OK, 1 row affected (0.04 sec)

mysql> select str_to_date(substr(dest,1,6),'%H%i%s') from sun;
+----------------------------------------+
| str_to_date(substr(dest,1,6),'%H%i%s') |
+----------------------------------------+
| NULL                                   |
+----------------------------------------+
1 row in set, 1 warning (0.00 sec)

mysql> select @@sql_mode;
+-------------------------------------------------------------------------------------------------------------------------------------------+
| @@sql_mode                                                                                                                                |
+-------------------------------------------------------------------------------------------------------------------------------------------+
| ONLY_FULL_GROUP_BY,STRICT_TRANS_TABLES,NO_ZERO_IN_DATE,NO_ZERO_DATE,ERROR_FOR_DIVISION_BY_ZERO,NO_AUTO_CREATE_USER,NO_ENGINE_SUBSTITUTION |
+-------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```
### 2. What did you expect to see? (Required)
str_to_date(substr(dest,1,6),'%H%i%s') returns `20:23:10` even the sql mode contains `NO_ZERO_IN_DATE,NO_ZERO_DATE`
### 3. What did you see instead (Required)
str_to_date(substr(dest,1,6),'%H%i%s') returns `NULL`
### 4. What is your TiDB version? (Required)
master@3765abbd434756b01212d7debf49406456620224

### PR DESCRIPTION
**Pull request by mengxin9014**
### What problem does this PR solve?

Issue Number: close #39146

Problem Summary:

### What is changed and how it works?
fix that str_to_date returns different result on no zero date sql mode
### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
fix that str_to_date returns different result on no zero date sql mode.
```

### CODE REVIEW

File: expression/builtin_time_vec.go [hunk 1]
```diff
@@ -1147,13 +1146,6 @@ func (b *builtinStrToDateDurationSig) vecEvalDuration(input *chunk.Chunk, result
 			result.SetNull(i, true)
 			continue
 		}
-		if hasNoZeroDateMode && (t.Year() == 0 || t.Month() == 0 || t.Day() == 0) {
-			if err := handleInvalidTimeError(b.ctx, types.ErrWrongValue.GenWithStackByArgs(types.DateTimeStr, t.String())); err != nil {
-				return err
-			}
-			result.SetNull(i, true)
-			continue
-		}
```
Comment (windtalker) [inline expression/builtin_time_vec.go:1156]: Looks like non-vector mode has already fix this?

Comment (mengxin9014) [reply expression/builtin_time_vec.go:1156]: yes. The non-vector mode ignores the value of hasNoZeroDateMode.

Comment (mengxin9014) [reply expression/builtin_time_vec.go:1156]: > Looks like non-vector mode has already fix this?

Yes, the non-vector mode ignores the value of hasNoZeroDateMode.

File: expression/integration_test.go [hunk 2]
```diff
@@ -7839,3 +7839,13 @@ func TestIfNullParamMarker(t *testing.T) {
...
 }
+
+func TestIssue39146(t *testing.T) {
+	store := testkit.CreateMockStore(t)
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("use test")
+	tk.MustExec("CREATE TABLE `sun` ( `dest` varchar(10) DEFAULT NULL ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;")
+	tk.MustExec("insert into sun values('20231020');")
+	tk.MustExec("set @@sql_mode = 'ONLY_FULL_GROUP_BY,STRICT_TRANS_TABLES,NO_ZERO_IN_DATE,NO_ZERO_DATE,ERROR_FOR_DIVISION_BY_ZERO,NO_AUTO_CREATE_USER,NO_ENGINE_SUBSTITUTION';")
+	tk.MustQuery(`select str_to_date(substr(dest,1,6),'%H%i%s') from sun;`).Check(testkit.Rows("20:23:10"))
```
[the remark this answers is not in the document]
Comment (mengxin9014) [reply expression/integration_test.go:7850]: > Please test both vector and non-vector execution.

done.

**Review by windtalker — APPROVED** [review]
LGTM

**Review by XuHuaiyu — APPROVED** [review]

**Comment by mengxin9014** [comment]
/needs-cherry-pick-release-6.2

**Pull request merged**